### PR TITLE
Add DOMJSEnv based on 'jsdom' in Node.js.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Review by @sjrd.

**Comment by nicolasstucki** [comment]
This PR will have merge conflicts with #2289 due to the changes in the CI matrix.

**Comment by nicolasstucki** [comment]
Rebased.

**Comment by nicolasstucki** [comment]
Oops, forgot to check if Jenkins had `jsdom` installed... it does not.

**Comment by nicolasstucki** [comment]
Jenkins, please retest this.

**Comment by nicolasstucki** [comment]
Jenkins, retest this please.

**Comment by nicolasstucki** [comment]
Note that I had to add some binary incompatibilities on the on the Node.js jsEnvs.

File: examples/testing/src/test/scala/DOMExistanceTest.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import scala.scalajs.js.Dynamic.global
+
+import org.scalajs.jasminetest.JasmineTest
+
+object DOMExistanceTest extends JasmineTest {
```
Comment (sjrd) [inline examples/testing/src/test/scala/DOMExistanceTest.scala]: Exist**e**nce with an 'e'. Existance is a pretty rare spelling. (Same in the file name, and in the HTML files.)

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala [hunk 2]
```diff
@@ -85,6 +85,30 @@ object ScalaJSPlugin extends AutoPlugin {
...
+     *  {{{
+     *  jsEnv := JSDOMNodeJSEnv().value
+     *  }}}
+     *
+     *  Note that the resulting [[sbt.Def.Setting Setting]] is not scoped at
+     *  all, but must be scoped in a project that has the ScalaJSPlugin enabled
+     *  to work properly.
+     *  Therefore, either put the upper line in your project settings (common
+     *  case) or scope it manually, using
+     *  [[sbt.ProjectExtra.inScope[* Project.inScope]].
```
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala]: This link looks weird. Does it work?

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 3]
```diff
@@ -569,7 +569,7 @@ object ScalaJSPluginInternal {
           val semantics = scalaJSLinker.value.semantics
           new RhinoJSEnv(semantics, withDOM = scalaJSRequestsDOM.value)
         } else if (scalaJSRequestsDOM.value) {
-          new PhantomJSEnv(jettyClassLoader = scalaJSPhantomJSClassLoader.value)
+          new JSDOMNodeJSEnv
```
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: It is probably too early for this change. We should give time to people to try out `JSDOMNodeJSEnv` opt-in before we change the default.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala [hunk 4]
```diff
@@ -79,244 +58,18 @@ class NodeJSEnv private (
...
-
-      comSocket != null
-    }
-
+      extends AsyncNodeRunner(libs, code) with NodeComJSRunner {
     override protected def initFiles(): Seq[VirtualJSFile] =
-      super.initFiles :+ comSetup
-
-    override protected def finalize(): Unit = close()
+      super.initFiles ++ comSetup
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: It seems both `ComDOMNodeRunner`and this class (`ComNodeRunner`) have the same implementation of `initFiles()`. Why is it not defined in `NodeComJSRunner`.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      }
+      Seq(new MemVirtualJSFile("codeWithJSDOMContext.js").withContent(jsDOMCode))
+    }
+
+    override protected def getJSFiles(): Seq[VirtualJSFile] =
+      initFiles() ++ codeWithJSDOMContext()
+
+    /** Libraries are loaded via scripts in Node.js */
+    override protected def getLibJSFiles(): Seq[VirtualJSFile] = {
+      customInitFiles() ++
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: What is the rationale for `customInitFiles()` to be in scripts but `initFiles()` to be top-level? In theory, they should both have the same status.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      libs.map(_.lib)
+    }
+
+    // Send code to Stdin
+    override protected def sendVMStdin(out: OutputStream): Unit = {
+      sendJS(getJSFiles(), out)
+    }
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: Both `AbstractDOMNodeRunner` and `AbstractBasicNodeRunner` define this method in the same way. Why is not defined in `AbstractNodeRunner` instead?

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  protected trait AbstractDOMNodeRunner extends AbstractNodeRunner {
+
+    protected def codeWithJSDOMContext(): Seq[VirtualJSFile] = {
+      val scriptsJSPaths = getLibJSFiles().map {
+        case file: FileVirtualFile => file.path
+        case file                  => libCache.materialize(file).getAbsolutePath
+      }
+      val scriptsStringPath = scriptsJSPaths.map('"' + _ + '"')
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: That is too weak. File names can contain weird characters. We have a function somewhere to properly escape JS strings (it's called `escapeJS`, I think).

Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: https://github.com/scala-js/scala-js/blob/ad4ba03003dd6b9a5e35279c22d57741d92bc9d5/ir/src/main/scala/org/scalajs/core/ir/Utils.scala#L54

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+           |        window["__ScalaJSEnv"] = __ScalaJSEnv;
+           |        window["scalajsCom"] = global.scalajsCom;
+           |        windowKeys = Object.keys(window);
+           |      } else {
+           |        console.log(error);
+           |      }
+           |    },
+           |    scripts: [${scriptsStringPath.mkString(", ")}],
+           |    onload: function (window) {
+           |      for (k in window) {
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: `var k`

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeComJSRunner.scala [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+import org.scalajs.core.tools.io._
+
+import org.scalajs.jsenv.Utils.OptDeadline
+import org.scalajs.jsenv._
+
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration._
+
+private[nodejs] trait NodeComJSRunner extends ComJSRunner {
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeComJSRunner.scala]: Why is this in its own file, and not as a `protected trait` in `AbstractNodeJSEnv`?

**Comment by sjrd** [comment]
That's all.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala (hunk 4 above)
Comment (nicolasstucki) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: To not use an abstract override.

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala (hunk 2 above)
Comment (nicolasstucki) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala]: The links does not work. It also fails for `NodeJSEnv` and `PhantomJSEnv`.

Comment (nicolasstucki) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPlugin.scala]: Actually all links to sbt `0.13.9` are broken, it looks like they are meant for sbt `0.12`. The links is resoved as: 
```
http://www.scala-sbt.org/0.13.9/api/index.html#sbt.ProjectExtra@inScope[T](scope:<?>,i:<?>):sbt.Def.Initialize[T]
```
and does not work.

But changing `0.13.9` to `0.12.1` makes the link work:
```
http://www.scala-sbt.org/0.12.1/api/index.html#sbt.ProjectExtra@inScope[T](scope:<?>,i:<?>):sbt.Def.Initialize[T]
```

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala (hunk 5 above)
Comment (nicolasstucki) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: No rationale.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeComJSRunner.scala (hunk 9 above)
Comment (nicolasstucki) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeComJSRunner.scala]: There is bug in the compiler. For some reason if do that it fails with the message `No getter for variable comSocket in trait AbstractNodeJSEnv$NodeComJSRunner` along with it's `scalac` tree.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala (hunk 6 above)
Comment (nicolasstucki) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: If I do that, `AbstractExtRunner.sendVMStdin` is chosen before `AbstractNodeRunner.sendVMStdin` due to the way the abstract classes and traits are mixed in.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala (hunk 4 above)
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: Maybe that would be a genuine use case for `abstract override`, then?

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala (hunk 6 above)
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala]: OK then maybe add a comment about that.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeComJSRunner.scala (hunk 9 above)
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeComJSRunner.scala]: Add a comment, then?

File: sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala [hunk 10]
```diff
@@ -22,8 +22,8 @@ import org.scalajs.core.tools.linker.backend.{LinkerBackend, OutputMode}
 
 import org.scalajs.jsenv._
 import org.scalajs.jsenv.rhino.RhinoJSEnv
-import org.scalajs.jsenv.nodejs.NodeJSEnv
-import org.scalajs.jsenv.phantomjs.{PhantomJSEnv, PhantomJettyClassLoader}
+import org.scalajs.jsenv.nodejs.{JSDOMNodeJSEnv, NodeJSEnv}
+import org.scalajs.jsenv.phantomjs.{PhantomJettyClassLoader, PhantomJSEnv}
```
Comment (sjrd) [inline sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala]: Useless diff.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala (hunk 4 above)
Comment (nicolasstucki) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: Yes, it looks like it.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala [hunk 11]
```diff
@@ -338,34 +83,12 @@ class NodeJSEnv private (
 
     // Send code to Stdin
     override protected def sendVMStdin(out: OutputStream): Unit = {
+     /* Do not factor this method out into AbstractNodeRunner or when mixin in
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: Oops, indentation issue.

File: project/BinaryIncompatibilities.scala [hunk 12]
```diff
@@ -9,6 +9,15 @@ object BinaryIncompatibilities {
...
   val JSEnvs = Seq(
+    ProblemFilters.exclude[MissingClassProblem]("org.scalajs.jsenv.nodejs.NodeJSEnv$AbstractNodeRunner"),
+    ProblemFilters.exclude[MissingTypesProblem]("org.scalajs.jsenv.nodejs.NodeJSEnv$NodeRunner"),
+    ProblemFilters.exclude[MissingTypesProblem]("org.scalajs.jsenv.nodejs.NodeJSEnv$AsyncNodeRunner"),
+    ProblemFilters.exclude[MissingTypesProblem]("org.scalajs.jsenv.nodejs.NodeJSEnv$ComNodeRunner"),
+    ProblemFilters.exclude[MissingMethodProblem]("org.scalajs.jsenv.nodejs.NodeJSEnv.acceptTimeout"),
+    ProblemFilters.exclude[MissingMethodProblem]("org.scalajs.jsenv.ComJSRunner.initFiles"),
+    ProblemFilters.exclude[MissingMethodProblem]("org.scalajs.jsenv.ComJSRunner.org$scalajs$jsenv$ComJSRunner$$super$initFiles"),
+    ProblemFilters.exclude[MissingMethodProblem]("org.scalajs.jsenv.ComJSRunner.comSetup"),
+    ProblemFilters.exclude[AbstractMethodProblem]("org.scalajs.jsenv.ExternalJSEnv.executable")
```
Comment (sjrd) [inline project/BinaryIncompatibilities.scala]: Wrap with 4 spaces after the `(`s.

Comment (sjrd) [inline project/BinaryIncompatibilities.scala]: Also 4 spaces wrt to `val JSEnvs` to begin with.

File: js-envs/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala [hunk 13]
```diff
@@ -115,7 +115,7 @@ class PhantomJSEnv(
     private[this] val recvBuf = mutable.Queue.empty[String]
     private[this] val fragmentsBuf = new StringBuilder
 
-    private def comSetup = {
+    override protected def comSetup(): Seq[VirtualJSFile] = {
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala]: Hum ... This should not have been made protected. It also does not make sense that this method be defined as high up as `ComJSRunner`.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala [hunk 14]
```diff
@@ -40,28 +35,7 @@ class NodeJSEnv private (
...
-      case t: ExternalJSEnv.NonZeroExitException =>
-        false
-    }
-  }
-
   protected def vmName: String = "Node.js"
-  protected def executable: String = nodejsPath
-
-  /** Retry-timeout to wait for the JS VM to connect */
-  protected val acceptTimeout = 5000
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: Removing this is not a good idea. It was specifically made protected and overridable at some point. This should be kept here, and that value should used, somehow, from `NodeComJSRunner`

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+import org.scalajs.core.tools.io._
+import org.scalajs.core.tools.jsdep.ResolvedJSDependency
+import org.scalajs.core.tools.logging.NullLogger
+import org.scalajs.jsenv._
+import org.scalajs.jsenv.Utils.OptDeadline
+
+import scala.concurrent.TimeoutException
+import scala.concurrent.duration._
+
+abstract class AbstractNodeJSEnv (nodejsPath: String, addArgs: Seq[String],
```
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: Spurious space before `(`

File: project/BinaryIncompatibilities.scala [hunk 16]
```diff
@@ -9,6 +9,18 @@ object BinaryIncompatibilities {
...
+      ProblemFilters.exclude[MissingClassProblem](
+          "org.scalajs.jsenv.nodejs.NodeJSEnv$AbstractNodeRunner"),
+      ProblemFilters.exclude[MissingTypesProblem](
+          "org.scalajs.jsenv.nodejs.NodeJSEnv$NodeRunner"),
+      ProblemFilters.exclude[MissingTypesProblem](
+          "org.scalajs.jsenv.nodejs.NodeJSEnv$AsyncNodeRunner"),
+      ProblemFilters.exclude[MissingTypesProblem](
+          "org.scalajs.jsenv.nodejs.NodeJSEnv$ComNodeRunner"),
+      ProblemFilters.exclude[MissingMethodProblem](
+          "org.scalajs.jsenv.nodejs.NodeJSEnv.acceptTimeout"),
```
Comment (sjrd) [inline project/BinaryIncompatibilities.scala]: This tipped me off.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala (hunk 14 above)
Comment (sjrd) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala]: (could be moved to `AbstractNodeJSEnv` instead, obviously)

File: project/BinaryIncompatibilities.scala (hunk 16 above)
Comment (nicolasstucki) [inline project/BinaryIncompatibilities.scala]: What?

**Comment by sjrd** [comment]
LGTM

**Comment by sjrd** [comment]
:tada:

**Pull request merged**